### ISSUE TITLE
Allow query of sensor names/BTADDR from DB on network

### ISSUE THREAD

**Issue by rlmalisz** [ThomDietrich/miflora-mqtt-daemon#35]
We'll be using these sensors to monitor outdoor plants, and because of range issues for BLE, will have multiple monitoring daemons.  It'd be lovely to be able to inform the daemons at wakeup what sensors they "own" without having to modify the config.ini by hand every time a new sensor gets added or a potted plant gets moved from one coverage area to another.

**Comment by ThomDietrich** [comment]
I get your use case but didn't understand which alternative you are seeking.

Retrieving data from a database sounds like out of scope of the daemon. As a tradeoff, wouldn't it be acceptable to modify the config file via a helper script before starting the daemon?

**Comment by rlmalisz** [comment]
It would be possible to stop/update/restart, but if it’s multiple machines, some of which are running on solar and only waking up to take measurements, it might be awkward.  I’m just arguing for a “soft” list of sensor devices.  The entry in the config.ini file could be a list as it is now, or alternately, an external entity to invoke to populate the list before each measurement pass.  It could return the list on stdout in the same format you use in the config file.   So while I agree that the actual query is out of scope, support for allowing such a query would be a nice thing.

—Richard

**Comment by ThomDietrich** [comment]
I see.

Well okay, I feel your idea is a bit too specific. I like the idea to centralize the configuration point for multiple sensors. How about the following:

We are already using MQTT for data publication. We could support an MQTT configuration topic per client, so that every client would subscript to a list of sensors he is supposed to query.

A few more details would need to be discussed but I like the idea in general.

...sadly I am currently not able to work on this. Would you like to push the idea forward?

**Comment by rlmalisz** [comment]
So how well does this work if the endpoint reading the sensors is asleep much of the time, and is not actively subscribed at the time an update gets published?  If this shows my ignorance of some of the nuances behind the pub/sub model MQTT supports, I apologize.

—Richard

**Comment by ThomDietrich** [comment]
That shouldn't be an issue. Tell me when you'd expect an update to take effect. Right now the config is only read on read during daemon startup. Would you expect it to be evaluated every time a new sensor readings update is scheduled?
That would totally be possible, we just have to figure out the details.

**Comment by rlmalisz** [comment]
I think only the sensor list would need re-evaluated, which is why I was envisioning it optionally being provided by some hook, oracle, whatever.  I was planning on getting readings every two hours, and should certainly be able to have the Pi Zero W wake up on a schedule that the central “provisioning daemon” would know about.  Then that pushes a minute or so after scheduled wakeup, the minions wait for the info and then scan.  I think, under that model, it’s fine if the mothership sends the sensor list on schedule irrespective of changes, at least, for the minions that spend a lot of time asleep.  The minions could publish back their next planned wakeup.

I think it’s doable.  I also think there’s still some merit in allowing the sensor list to be provided by an arbitrary helper as listed in the config file.  Consider the (truly) degenerate case: the helper, when invoked, could simply do an lescan for a minute or two, and return the list with BTADDRS as both name and BTADDR.  All the subscriber to the published data needs to know is the association of the BTADDR to the sensor and its plant or whatever.  I move a pot from one area of coverage to another, and it just takes care of itself.  The list could be a network-mounted file, and the helper just cats it.

I’m not suggesting making the daemon a lot more complicated.  I’m just suggesting having the option of having the sensor list acquired by opening a pipe to some process or helper, instead of hard-coded.  If the user chooses to manage that via MQTT and work out the timing, plumbing, etc, that’s fine.  I’m in favor of delegating this to the user…all we need is to allow the hook, to invoke it before each set of readings if provided, and make sure the format the hook responds with is well-understood.

—Richard

**Comment by rlmalisz** [comment]
Retained messages make this all pretty straightforward.  I was not aware of them being part of the MQTT protocol.  As I said before, I am new to MQTT.

—Richard

**Comment by ThomDietrich** [comment]
#39

**Comment by ThomDietrich** [comment]
@rlmalisz How is your fork working for you? https://github.com/rlmalisz/miflora-mqtt-daemon/commit/b689503926e4af20fa86d63f4256e59fda6370bf

**Comment by ThomDietrich** [comment]
@bangom this ticket is a bit older but still very relevant. Are you also interested in this? https://github.com/ThomDietrich/miflora-mqtt-daemon/issues/35#issuecomment-382558987

**Comment by rlmalisz** [comment]
I have no complaints about my fork.  We have a half-dozen "readers" around the house and yard.  Not too many sensors deployed this time of year, but in summer, a couple dozen.  Readers collect their assigned sensor readings every three hours.  Just before that interval, a cron (OSX, so actually launchctl) script queries a MySQL DB for assignments, and publishes them to MQTT.  Readers' scripts run, get their assignments, and gather readings, and publish them.  Another OSX scripts runs after those finish, and collects published readings and adds them to MySQL.  It also posts warnings about any readings that are too low (or high) for sensors.  Have a PHP script that uses Google Charts to display recent data queried from the DB.  Image attached.  But adding/deleting/moving a sensor is a simple update to a couple DB tables.  This was what I wanted, so am quite happy.
![Screen Shot 2020-03-29 at 9 52 46 AM](https://user-images.githubusercontent.com/1276500/77855089-230ec580-71a3-11ea-894b-9c8976bd330d.png)

**Comment by bangom** [comment]
> @bangom this ticket is a bit older but still very relevant. Are you also interested in this? [#35 (comment)](https://github.com/ThomDietrich/miflora-mqtt-daemon/issues/35#issuecomment-382558987)

Most definitely. Next miflora-mqtt-daemon development step should be implementing distributed miflora-mqtt-daemons configuration from single place (master node?). This is IMHO primary usecase for miflora-mqtt-daemon (multiple btle2wifi clusters of RPI0 connected to central hub like Home Assistant). Let me think this over and get back with my ideas :) But from the initial analysis of this issue I like the idea of using MQTT (sensor data and daemons configuration updates).

**Comment by ThomDietrich** [comment]
I already have an idea. Let me post a proposal this evening
